executor/union_scan.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/pingcap/tidb/util/types"
+)
+
+type dirtyDB struct {
+	tables map[int64]*dirtyTable
+}
+
+func (udb *dirtyDB) AddRow(tid, handle int64, row []types.Datum) {
+	dt := udb.getDirtyTable(tid)
+	for i := range row {
+		if row[i].Kind() == types.KindString {
+			row[i].SetBytes(row[i].GetBytes())
+		}
+	}
+	dt.addedRows[handle] = row
+}
+
+func (udb *dirtyDB) DeleteRow(tid int64, handle int64) {
+	dt := udb.getDirtyTable(tid)
+	delete(dt.addedRows, handle)
+	dt.deletedRows[handle] = struct{}{}
+}
+
+func (udb *dirtyDB) TruncateTable(tid int64) {
+	dt := udb.getDirtyTable(tid)
+	dt.addedRows = make(map[int64][]types.Datum)
+	dt.truncated = true
+}
+
+func (udb *dirtyDB) getDirtyTable(tid int64) *dirtyTable {
+	dt, ok := udb.tables[tid]
+	if !ok {
+		dt = &dirtyTable{
+			addedRows:   make(map[int64][]types.Datum),
+			deletedRows: make(map[int64]struct{}),
+		}
+		udb.tables[tid] = dt
+	}
+	return dt
+}
+
+type dirtyTable struct {
+	// key is handle.
+	addedRows   map[int64][]types.Datum
+	deletedRows map[int64]struct{}
+	truncated   bool
+}
+
+type dirtyDBKeyType int
+
+func (u dirtyDBKeyType) String() string {
+	return "dirtyDBKeyType"
+}
+
+// DirtyDBKey is the key to *dirtyDB for a context.
+const DirtyDBKey dirtyDBKeyType = 1
+
+func getDirtyDB(ctx context.Context) *dirtyDB {
+	var udb *dirtyDB
+	x := ctx.Value(DirtyDBKey)
+	if x == nil {
+		udb = &dirtyDB{tables: make(map[int64]*dirtyTable)}
+		ctx.SetValue(DirtyDBKey, udb)
+	} else {
+		udb = x.(*dirtyDB)
+	}
+	return udb
+}
+
+// UnionScanExec merges the rows from dirty table and the rows from XAPI request.
+type UnionScanExec struct {
+	ctx   context.Context
+	Src   Executor
+	dirty *dirtyTable
+	// srcUsedIndex is the column offsets of the index which Src executor has used.
+	usedIndex []int
+	desc      bool
+	condition ast.ExprNode
+
+	addedRows []*Row
+	cursor    int
+	sortErr   error
+}
+
+// Fields implements Executor Fields interface.
+func (us *UnionScanExec) Fields() []*ast.ResultField {
+	return us.Src.Fields()
+}
+
+// Next implements Execution Next interface.
+func (us *UnionScanExec) Next() (*Row, error) {
+	for {
+		snapshotRow, err := us.Src.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		var addedRow *Row
+		if us.cursor < len(us.addedRows) {
+			addedRow = us.addedRows[us.cursor]
+		}
+		if us.dirty.truncated || snapshotRow == nil {
+			if addedRow != nil {
+				for i, field := range us.Src.Fields() {
+					field.Expr.SetDatum(addedRow.Data[i])
+				}
+				us.cursor++
+				return addedRow, nil
+			}
+			return nil, nil
+		}
+		if addedRow == nil {
+			return snapshotRow, nil
+		}
+		if len(snapshotRow.RowKeys) != 1 {
+			return nil, ErrRowKeyCount
+		}
+		snapshotHandle := snapshotRow.RowKeys[0].Handle
+		if _, ok := us.dirty.deletedRows[snapshotHandle]; ok {
+			continue
+		}
+		if _, ok := us.dirty.addedRows[snapshotHandle]; ok {
+			// If src handle appears in added rows, it means there is conflict and the transaction will fail to
```

Comment:
We first insert a row, with an assumption that there is no conflict, then we select rows from snapshot, found that the row exists.